File: 2395-longest-binary-subsequence-less-than-or-equal-to-k/longest-binary-subsequence-less-than-or-equal-to-k.java
class Solution {
    public int longestSubsequence(String s, int k) {
        int res = 0, power = 1;
        char[] arr = s.toCharArray();
        for (int i=arr.length-1; i>=0; i--) {
            if (arr[i] == '0') {
                res++;
            } else if (power <= k) {
                k -= power;
                res++;
            }
            if (power <= k) power *= 2;
        }
        return res;
    }
}

File: 2395-longest-binary-subsequence-less-than-or-equal-to-k/longest-binary-subsequence-less-than-or-equal-to-k.py
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        power = 1
        res = 0
        for val in s[::-1]:
            if val == '0':
                res += 1
            elif val == '1' and power <= k:
                k -= power
                res += 1
            if power <= k:
                power *= 2

        return res